### PR TITLE
PS-5747 : Remove AHI overhead for DDLs when AHI is disabled

### PR DESCRIPTION
**Pull request by satya-bodapati**
Problem:
--------
Even when AHI is disabled, there is overhead because of AHI.
This is because of checks required for AHI to delete hash entries
of pages being dropped.

These checks are not necessary if we know for sure that AHI is already
disabled.

Fix:
----
1. Take extra dict_sys mutex when enabling AHI (btr_search_enable()). This
   is now similar to AHI disable (btr_search_disable())

2. btr_search_enabled variable read under dict_sys mutex is now accurate
   and can be used to know if AHI is disabled for sure and can't be enabled
   concurrently.

Fix is similar to PS-5576 - Server stalls because ALTER TABLE on partitioned
                             table holds dict mutex

This fix is more generic and should cover DROP TABLE, DROP INDEX also (apart
from TRUNCATE)

### CODE REVIEW

File: storage/innobase/btr/btr0btr.cc [hunk 1]
```diff
@@ -1192,17 +1192,26 @@ btr_create(
 /** Free a B-tree except the root page. The root page MUST be freed after
 this by calling btr_free_root.
 @param[in,out]	block		root page
-@param[in]	log_mode	mtr logging mode */
+@param[in]	log_mode	mtr logging mode
+@param[in]	is_ahi_allowed	false for intrinsic tables else true */
```
Comment (laurynas-biveinis) [inline storage/innobase/btr/btr0btr.cc]: argument name and description differ. Perhaps the description explains when AHI is allowed, but that should be made clear

File: storage/innobase/btr/btr0sea.cc [hunk 2]
```diff
@@ -429,16 +429,28 @@ btr_search_disable(
 	btr_search_mem_accounting_validate();
 }
 
-/** Enable the adaptive hash search system. */
+/** Enable the adaptive hash search system
+@param[in]	need_mutex	need to acquire dict_sys->mutex */
```
Comment (laurynas-biveinis) [inline storage/innobase/btr/btr0sea.cc]: Add that the mutex must be already held if false

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Nice. Minor comments only

File: storage/innobase/btr/btr0btr.cc (hunk 1 above)
Comment (satya-bodapati) [reply storage/innobase/btr/btr0btr.cc]: ahi is not applicable/allowed for intrinsic tables, For non-intrinsic tables, AHI is applicable/allowed.

Didn't get the question. What should I change here?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
As discussed on Slack

**Comment by satya-bodapati** [comment]
Ready for review

**Review by laurynas-biveinis — APPROVED** [review]

**Comment by satya-bodapati** [comment]
I will submit a 8.0 version patch too. It will be done via cherry-pick( This is because PS-5576 was done on latest 5.7)

**Pull request merged**